rust/xaynet-sdk/src/state_machine/io.rs

```diff
@@ -0,0 +1,179 @@
+use std::error::Error;
+
+use xaynet_core::{
+    common::RoundParameters,
+    mask::Model,
+    SumDict,
+    SumParticipantPublicKey,
+    UpdateSeedDict,
+};
+
+use crate::{ModelStore, Notify, XaynetClient};
+
+pub(crate) fn boxed_io<X, M, N>(
+    xaynet_client: X,
+    model_store: M,
+    notifier: N,
+) -> Box<dyn IO<Model = Box<dyn AsRef<Model> + Send>>>
+where
+    X: XaynetClient + Send + 'static,
+    M: ModelStore + Send + 'static,
+    N: Notify + Send + 'static,
+{
+    Box::new((xaynet_client, model_store, notifier))
+}
+
+pub struct DefaultNotifier;
+impl Notify for DefaultNotifier {}
+
+#[async_trait]
+pub(crate) trait IO: Send + 'static {
+    type Model;
+
+    async fn load_model(&mut self) -> Result<Option<Self::Model>, Box<dyn Error>>;
+    async fn get_round_params(&mut self) -> Result<RoundParameters, Box<dyn Error>>;
+    async fn get_sums(&mut self) -> Result<Option<SumDict>, Box<dyn Error>>;
+    async fn get_seeds(
+        &mut self,
+        pk: SumParticipantPublicKey,
+    ) -> Result<Option<UpdateSeedDict>, Box<dyn Error>>;
+    async fn get_mask_length(&mut self) -> Result<Option<u64>, Box<dyn Error>>;
+    async fn get_model(&mut self) -> Result<Option<Model>, Box<dyn Error>>;
+    async fn send_message(&mut self, msg: Vec<u8>) -> Result<(), Box<dyn Error>>;
+
+    fn notify_new_round(&mut self);
+    fn notify_sum(&mut self);
+    fn notify_update(&mut self);
+    fn notify_idle(&mut self);
+}
+
+#[async_trait]
+impl<X, M, N> IO for (X, M, N)
```

Comment:
i'd suggest to have a struct with named fields instead of the triple to improve readability.

Reply:
done